File: rad/engine/05_exec/generative/recursive.go
package generative

// Recursive-query synthesis: a random directed graph plus a correct-by-
// construction recursive query over it, for the three-way differential. Think
// like a compiler team, not a SQL one — generate graphs and traversals, then
// hold the executor to the reference interpreter.
//
// Termination is guaranteed by construction so every case checks a *result*,
// never the iteration cap (cycles and caps are covered by hand-written tests):
//
//   - An acyclic graph (edges only point forward in node order) terminates
//     under either accumulation mode with any recursive state, so those cases
//     carry the rich state shapes.
//   - An arbitrary graph may contain cycles, so those cases use `accumulation:
//     new` over id-only rows, which closes on the finite node set.

import (
	"fmt"

	"pgregory.net/rapid"

	"github.com/Southclaws/rad/rad/engine/02_catalog/model"
	lir "github.com/Southclaws/rad/rad/engine/03_lir"
)

// GraphCatalog is the fixed schema the recursive generator traverses: a
// directed edge relation whose `w` (weight) and nullable `note` feed recursive
// state. The index on `src` gives the planner an access path, so the forced-
// full-scan leg of the differential is a genuine alternative.
func GraphCatalog() *Catalog {
	return &Catalog{Tables: []Table{{
		Name: "edges",
		Columns: []Column{
			{Name: "src", Type: model.TypeText},
			{Name: "dst", Type: model.TypeText},
			{Name: "w", Type: model.TypeInt64},
			{Name: "note", Type: model.TypeText, Nullable: true},
		},
		PrimaryKey: []string{"src", "dst"},
		Indexes:    [][]string{{"src"}},
	}}}
}

// GraphEdge is one directed edge with its state-feeding attributes; a nil Note
// is a NULL.
type GraphEdge struct {
	Src, Dst string
	W        int64
	Note     *string
}

// Graph is a generated directed graph. Acyclic records how it was drawn: when
// true, every edge points strictly forward in node order, so the query may
// carry rich state under either accumulation mode.
type Graph struct {
	Nodes   []string
	Edges   []GraphEdge
	Acyclic bool
}

// GenGraph draws a random directed graph: 2..6 nodes and up to 2n distinct
// edges. An acyclic draw restricts edges to forward pairs (i < j); an
// unrestricted draw admits back-edges and self-loops, hence cycles.
func GenGraph(rt *rapid.T) Graph {
	n := rapid.IntRange(2, 6).Draw(rt, "graph_nodes")
	nodes := make([]string, n)
	for i := range nodes {
		nodes[i] = fmt.Sprintf("n%d", i)
	}

	acyclic := rapid.Bool().Draw(rt, "graph_acyclic")
	m := rapid.IntRange(0, 2*n).Draw(rt, "graph_edges")
	seen := map[[2]int]bool{}
	var edges []GraphEdge
	for range m {
		i := rapid.IntRange(0, n-1).Draw(rt, "edge_src")
		j := rapid.IntRange(0, n-1).Draw(rt, "edge_dst")
		if acyclic {
			if i == j {
				continue
			}
			if i > j {
				i, j = j, i
			}
		}
		if seen[[2]int{i, j}] {
			continue
		}
		seen[[2]int{i, j}] = true
		edges = append(edges, GraphEdge{
			Src:  nodes[i],
			Dst:  nodes[j],
			W:    int64(rapid.IntRange(0, 5).Draw(rt, "edge_w")),
			Note: genNote(rt),
		})
	}
	return Graph{Nodes: nodes, Edges: edges, Acyclic: acyclic}
}

// genNote draws an edge note: NULL a third of the time, else a small value.
func genNote(rt *rapid.T) *string {
	if rapid.IntRange(0, 2).Draw(rt, "edge_note_null") == 0 {
		return nil
	}
	s := rapid.SampledFrom([]string{"x", "y", ""}).Draw(rt, "edge_note")
	return &s
}

// GraphData renders a graph's edges as rows for the edges table, keyed by
// table name for the interpreter's scan and the engine's inserts.
func GraphData(g Graph) map[string][]lir.Row {
	rows := make([]lir.Row, len(g.Edges))
	for i, e := range g.Edges {
		note := lir.Null(model.TypeText)
		if e.Note != nil {
			note = lir.Text(*e.Note)
		}
		rows[i] = lir.Row{"src": lir.Text(e.Src), "dst": lir.Text(e.Dst), "w": lir.Int64(e.W), "note": note}
	}
	return map[string][]lir.Row{"edges": rows}
}

// stateCol is one recursive state column: its declared type, the anchor's
// initial cell (nil = NULL), and the step expression computing it from the
// frontier (scope "p") and the joined edge (scope "e").
type stateCol struct {
	name     string
	typ      model.Type
	nullable bool
	initial  any
	step     func() lir.Expr
}

var (
	// depth: a hop counter — parent.depth + 1.
	depthCol = stateCol{
		name: "depth", typ: model.TypeInt64, initial: int64(0),
		step: func() lir.Expr { return lir.Binary{Op: lir.OpAdd, L: qcol("p", "depth"), R: qlit(int64(1))} },
	}
	// cost: an accumulated edge weight — parent.cost + edge.w.
	costCol = stateCol{
		name: "cost", typ: model.TypeInt64, initial: int64(0),
		step: func() lir.Expr { return lir.Binary{Op: lir.OpAdd, L: qcol("p", "cost"), R: qcol("e", "w")} },
	}
	// tag / tag2: a nullable text carried from the reaching edge's note,
	// initialised NULL — exercises nullability reconciliation and NULL identity.
	tagCol  = stateCol{name: "tag", typ: model.TypeText, nullable: true, step: func() lir.Expr { return qcol("e", "note") }}
	tag2Col = stateCol{name: "tag2", typ: model.TypeText, nullable: true, step: func() lir.Expr { return qcol("e", "note") }}
	// flag: a boolean that flips each hop — not(parent.flag).
	flagCol = stateCol{
		name: "flag", typ: model.TypeBool, initial: true,
		step: func() lir.Expr { return lir.Unary{Op: lir.OpNot, X: qcol("p", "flag")} },
	}
	// num: a nullable int carried from the reaching edge's weight, initialised
	// NULL — a nullable int alongside a nullable text stresses the signature.
	numCol = stateCol{name: "num", typ: model.TypeInt64, nullable: true, step: func() lir.Expr { return qcol("e", "w") }}
)

// stateShapes are the recursive-state signatures the generator draws from (id
// is always present and implicit). Ranging from none through multiple nullable
// columns, they exercise recursive typing as much as traversal.
var stateShapes = [][]stateCol{
	{},
	{depthCol},
	{costCol},
	{tagCol},
	{depthCol, flagCol},
	{tagCol, numCol},
	{depthCol, costCol, tagCol, flagCol, tag2Col},
}

// GenRecursiveQuery synthesises a reachability query over g: a random anchor
// set, an accumulation mode, and a recursive-state signature. Cyclic graphs
// are constrained to id-only rows under `accumulation: new` so the fixpoint
// always closes; acyclic graphs carry any state under either mode. The query
// is well-formed by construction — one frontier reference in the step, a
// monotone join, an anchor free of self-reference.
func GenRecursiveQuery(rt *rapid.T, g Graph) lir.Query {
	nRoots := rapid.IntRange(1, min(3, len(g.Nodes))).Draw(rt, "root_count")
	roots := rapid.Permutation(g.Nodes).Draw(rt, "root_perm")[:nRoots]

	accumulation := lir.AccumulateNew
	shape := stateShapes[0]
	if g.Acyclic {
		if rapid.Bool().Draw(rt, "accumulate_all") {
			accumulation = lir.AccumulateAll
		}
		shape = rapid.SampledFrom(stateShapes).Draw(rt, "state_shape")
	}

	anchorCols := []lir.RowsCol{{Name: "id", Kind: lir.KindText}}
	for _, sc := range shape {
		anchorCols = append(anchorCols, lir.RowsCol{Name: sc.name, Kind: lir.Kind(string(sc.typ)), Nullable: sc.nullable})
	}
	anchorVals := make([][]any, len(roots))
	for i, r := range roots {
		row := []any{r}
		for _, sc := range shape {
			row = append(row, sc.initial)
		}
		anchorVals[i] = row
	}
	anchor := lir.Rows{Scope: "a", Columns: anchorCols, Values: anchorVals}

	stepFields := []lir.ProjField{{As: "id", Expr: qcol("e", "dst")}}
	for _, sc := range shape {
		stepFields = append(stepFields, lir.ProjField{As: sc.name, Expr: sc.step()})
	}
	step := lir.Project{
		Input: lir.Join{
			Left:  qscan("edges", "e"),
			Right: lir.RecursiveRef{Binding: "rec", Scope: "p"},
			Kind:  lir.InnerJoin,
			On:    qeq(qcol("e", "src"), qcol("p", "id")),
		},
		Scope:  "s",
		Fields: stepFields,
	}

	return lir.Query{
		Card:     lir.CardMany,
		Bindings: map[string]lir.Relation{"rec": lir.Recursive{Anchor: anchor, Step: step, Accumulation: accumulation}},
		Root: lir.Order{
			Input: lir.Ref{Binding: "rec", Scope: "r"},
			Terms: []lir.OrderTerm{{Expr: qlit(true)}},
		},
	}
}

File: tests/generative/recursive.rs
use std::collections::{BTreeMap, HashMap, HashSet};

use rad::engine::catalog::identity::SchemaId;
use rad::engine::catalog::model::{ColumnDef, IndexDef, ScalarType, Schema, TableDef};
use rad::engine::lir::{
    BinaryOp, Expr, JoinKind, Kind, Literal, OrderTerm, ProjectField, Query, RawScalar,
    RecursiveAccumulation, Relation, RootCardinality, RowsColumn, UnaryOp, Value,
};

use super::{Case, CaseKind, Choices, decisions_from_seed};

pub fn recursive_case(seed: u64) -> Case {
    generate(decisions_from_seed(seed))
}

pub fn generate(decisions: Vec<u64>) -> Case {
    let mut choices = Choices::new(&decisions);
    let catalog = graph_catalog();
    let node_count = choices.range(2, 6);
    let nodes = (0..node_count)
        .map(|index| format!("n{index}"))
        .collect::<Vec<_>>();
    let acyclic = choices.coin();
    let edge_attempts = choices.range(0, node_count * 2);
    let mut seen = HashSet::new();
    let mut rows = Vec::new();
    for _ in 0..edge_attempts {
        let mut source = choices.index(node_count);
        let mut destination = choices.index(node_count);
        if acyclic {
            if source == destination {
                continue;
            }
            if source > destination {
                std::mem::swap(&mut source, &mut destination);
            }
        }
        if !seen.insert((source, destination)) {
            continue;
        }
        rows.push(HashMap::from([
            ("src".into(), Value::Text(nodes[source].clone())),
            ("dst".into(), Value::Text(nodes[destination].clone())),
            ("w".into(), Value::Int64(choices.range(0, 5) as i64)),
            (
                "note".into(),
                if choices.chance(3) {
                    Value::Null(ScalarType::Text)
                } else {
                    Value::Text(["", "x", "y"][choices.index(3)].into())
                },
            ),
        ]));
    }
    let data = BTreeMap::from([("edges".into(), rows)]);
    let query = recursive_query(&nodes, acyclic, &mut choices);
    Case {
        kind: CaseKind::Recursive,
        decisions,
        catalog,
        data,
        query,
        ordered: false,
    }
}

fn graph_catalog() -> Schema {
    Schema::from_definitions(vec![TableDef {
        id: id(1),
        name: "edges".into(),
        columns: vec![
            column(1, "src", ScalarType::Text, false),
            column(2, "dst", ScalarType::Text, false),
            column(3, "w", ScalarType::Int64, false),
            column(4, "note", ScalarType::Text, true),
        ],
        primary_key: vec!["src".into(), "dst".into()],
        indexes: vec![IndexDef {
            name: "edges_src_idx".into(),
            columns: vec!["src".into()],
            unique: false,
        }],
        foreign_keys: Vec::new(),
    }])
}

fn recursive_query(nodes: &[String], acyclic: bool, choices: &mut Choices<'_>) -> Query {
    let root_count = choices.range(1, nodes.len().min(3));
    let mut roots = nodes.to_vec();
    for index in 0..roots.len() {
        let other = index + choices.index(roots.len() - index);
        roots.swap(index, other);
    }
    roots.truncate(root_count);

    let shape = if acyclic { choices.index(7) } else { 0 };
    let accumulation = if acyclic && choices.coin() {
        RecursiveAccumulation::All
    } else {
        RecursiveAccumulation::New
    };
    let states = state_shape(shape);
    let mut columns = vec![RowsColumn {
        name: "id".into(),
        kind: Kind::Text,
        nullable: false,
    }];
    columns.extend(states.iter().map(|state| RowsColumn {
        name: state.name.into(),
        kind: state.kind,
        nullable: state.nullable,
    }));
    let values = roots
        .into_iter()
        .map(|root| {
            let mut row = vec![RawScalar::Text(root)];
            row.extend(states.iter().map(|state| state.initial.clone()));
            row
        })
        .collect();
    let anchor = Relation::Rows {
        scope: "a".into(),
        columns,
        values,
    };

    let mut fields = vec![ProjectField {
        name: "id".into(),
        expression: column_expr("e", "dst"),
    }];
    fields.extend(states.iter().map(|state| ProjectField {
        name: state.name.into(),
        expression: (state.step)(),
    }));
    let step = Relation::Project {
        input: Box::new(Relation::Join {
            left: Box::new(Relation::Scan {
                table: "edges".into(),
                scope: "e".into(),
            }),
            right: Box::new(Relation::RecursiveRef {
                binding: "rec".into(),
                scope: "p".into(),
            }),
            kind: JoinKind::Inner,
            on: Expr::Binary {
                op: BinaryOp::Eq,
                left: Box::new(column_expr("e", "src")),
                right: Box::new(column_expr("p", "id")),
            },
        }),
        scope: Some("s".into()),
        spread: Vec::new(),
        fields,
    };
    Query {
        root: Relation::Order {
            input: Box::new(Relation::Ref {
                binding: "rec".into(),
                scope: "r".into(),
            }),
            terms: vec![OrderTerm {
                expression: literal_bool(true),
                descending: false,
            }],
        },
        cardinality: RootCardinality::Many,
        bindings: HashMap::from([(
            "rec".into(),
            Relation::Recursive {
                anchor: Box::new(anchor),
                step: Box::new(step),
                accumulation,
            },
        )]),
    }
}

struct State {
    name: &'static str,
    kind: Kind,
    nullable: bool,
    initial: RawScalar,
    step: fn() -> Expr,
}

fn state_shape(shape: usize) -> Vec<State> {
    let mut states = Vec::new();
    for name in match shape {
        0 => &[][..],
        1 => &["depth"][..],
        2 => &["cost"][..],
        3 => &["tag"][..],
        4 => &["depth", "flag"][..],
        5 => &["tag", "num"][..],
        _ => &["depth", "cost", "tag", "flag", "tag2"][..],
    } {
        states.push(match *name {
            "depth" => State {
                name,
                kind: Kind::Int64,
                nullable: false,
                initial: RawScalar::Number("0".into()),
                step: depth,
            },
            "cost" => State {
                name,
                kind: Kind::Int64,
                nullable: false,
                initial: RawScalar::Number("0".into()),
                step: cost,
            },
            "tag" | "tag2" => State {
                name,
                kind: Kind::Text,
                nullable: true,
                initial: RawScalar::Null,
                step: note,
            },
            "flag" => State {
                name,
                kind: Kind::Bool,
                nullable: false,
                initial: RawScalar::Bool(true),
                step: flag,
            },
            "num" => State {
                name,
                kind: Kind::Int64,
                nullable: true,
                initial: RawScalar::Null,
                step: weight,
            },
            _ => unreachable!(),
        });
    }
    states
}

fn depth() -> Expr {
    add(column_expr("p", "depth"), literal_int(1))
}
fn cost() -> Expr {
    add(column_expr("p", "cost"), column_expr("e", "w"))
}
fn note() -> Expr {
    column_expr("e", "note")
}
fn weight() -> Expr {
    column_expr("e", "w")
}
fn flag() -> Expr {
    Expr::Unary {
        op: UnaryOp::Not,
        expression: Box::new(column_expr("p", "flag")),
    }
}
fn add(left: Expr, right: Expr) -> Expr {
    Expr::Binary {
        op: BinaryOp::Add,
        left: Box::new(left),
        right: Box::new(right),
    }
}
fn column_expr(scope: &str, name: &str) -> Expr {
    Expr::Column {
        scope: scope.into(),
        name: name.into(),
    }
}
fn literal_int(value: i64) -> Expr {
    Expr::Literal(Literal {
        raw: RawScalar::Number(value.to_string()),
        kind: Some(Kind::Int64),
    })
}
fn literal_bool(value: bool) -> Expr {
    Expr::Literal(Literal {
        raw: RawScalar::Bool(value),
        kind: Some(Kind::Bool),
    })
}
fn column(id_value: u32, name: &str, scalar_type: ScalarType, nullable: bool) -> ColumnDef {
    ColumnDef {
        id: id(id_value),
        name: name.into(),
        scalar_type,
        nullable,
        format: String::new(),
        default: None,
    }
}
fn id(value: u32) -> SchemaId {
    SchemaId::new(value).unwrap()
}
